File: src/60-days-02/day-061/koko-eating-bananas.kt
class Solution {
  fun minEatingSpeed(piles: IntArray, h: Int): Int {    
    var low: Int = 1
    var high: Int = piles.max() ?: 0
    while(low < high){
      val mid = low + (high - low) / 2
      if(isValid(mid, piles, h)){
        high = mid
        continue
      }
      low = mid + 1
    }
    return low
  }
    
  fun isValid(maxPiles: Int, piles: IntArray, h: Int): Boolean {
    var hours: Int = 0
    for(i: Int in piles){
      val time = i / maxPiles
      hours += time
      if(i % maxPiles != 0) hours++
    }
    if (hours <= h) return true
    return false
  }

}

File: src/60-days-02/day-061/koko-eating-bananas.dart
import 'dart:math';

class Solution {
  int minEatingSpeed(List<int> piles, int h) {
    int low = 1;
    int high = piles.reduce((curr, next) => curr > next ? curr : next);
    while (low < high) {
      int mid = low + ((high - low) / 2).floor();
      if (isValid(mid, piles, h)) {
        high = mid;
        continue;
      }
      low = mid + 1;
    }
    return low;
  }

  bool isValid(int maxpiles, List<int> piles, int h) {
    int hours = 0;
    for (int i in piles) {
      int time = (i / maxpiles).floor();
      hours += time;
      if (i % maxpiles != 0) hours++;
    }
    if (hours <= h) return true;
    return false;
  }
}
